File: backend/src/main/java/com/skillex/service/MessageService.java
package com.skillex.service;

import com.skillex.dto.common.PagedResponse;
import com.skillex.dto.message.ConversationDto;
import com.skillex.dto.message.MessageDto;

import java.util.List;

/**
 * Contract for direct peer-to-peer messaging.
 */
public interface MessageService {

    /** Returns all conversations for the given user, sorted by most-recent message. */
    List<ConversationDto> getConversations(String userId);

    /** Paginated message history between userId and peerId. */
    PagedResponse<MessageDto> getHistory(String userId, String peerId, int page, int size);

    /** Persists a new message and returns the saved DTO. */
    MessageDto sendMessage(String senderId, String receiverId, String content, String type, String imageUrl);

    /** Marks all messages FROM peerId TO userId as read. */
    void markRead(String userId, String peerId);
}

File: frontend/src/services/messageService.ts
import { api } from './api';

export interface MessageDto {
  id: string;
  senderId: string;
  receiverId: string;
  content: string;
  /** "TEXT" | "IMAGE" */
  type: string;
  imageUrl?: string | null;
  isRead: boolean;
  createdAt: string;
}

export interface ConversationDto {
  peerId: string;
  peerName: string;
  peerAvatar?: string | null;
  peerUniversity?: string | null;
  peerIsOnline: boolean;
  lastMessage: string;
  lastMessageType: string;
  lastMessageTime?: string | null;
  unreadCount: number;
}

export interface PagedMessages {
  content: MessageDto[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  last: boolean;
}

export const MessageService = {
  /** GET /api/messages/conversations — all conversations, most-recent first */
  getConversations: (): Promise<ConversationDto[]> =>
    api.get<ConversationDto[]>('/messages/conversations'),

  /**
   * GET /api/messages/{peerId}?page=0&size=50
   * Paginated message history between the current user and a peer.
   */
  getHistory: (peerId: string, page = 0, size = 50): Promise<PagedMessages> =>
    api.get<PagedMessages>(`/messages/${peerId}?page=${page}&size=${size}`),

  /** PATCH /api/messages/{peerId}/read — mark all messages from peerId as read */
  markRead: (peerId: string): Promise<void> =>
    api.patch<void>(`/messages/${peerId}/read`, {}),
};
